### PR TITLE
Guard against unwinding in cleanup code

### PR DESCRIPTION
**Pull request by nbdd0121**
Currently the only safe guard we have against double unwind is the panic count (which is local to Rust). When double unwinds indeed happen (e.g. C++ exception + Rust panic, or two C++ exceptions), then the second unwind actually goes through and the first unwind is leaked. This can cause UB. cc rust-lang/project-ffi-unwind#6

E.g. given the following C++ code:
```c++
extern "C" void foo() {
    throw "A";
}

extern "C" void execute(void (*fn)()) {
    try {
        fn();
    } catch(...) {
    }
}
```

This program is well-defined to terminate:
```c++
struct dtor {
    ~dtor() noexcept(false) {
        foo();
    }
};

void a() {
    dtor a;
    dtor b;
}

int main() {
    execute(a);
    return 0;
}
```

But this Rust code doesn't catch the double unwind:
```rust
extern "C-unwind" {
    fn foo();
    fn execute(f: unsafe extern "C-unwind" fn());
}

struct Dtor;

impl Drop for Dtor {
    fn drop(&mut self) {
        unsafe { foo(); }
    }
}

extern "C-unwind" fn a() {
    let _a = Dtor;
    let _b = Dtor;
}

fn main() {
    unsafe { execute(a) };
}
```

To address this issue, this PR adds an unwind edge to an abort block, so that the Rust example aborts. This is similar to how clang guards against double unwind (except clang calls terminate per C++ spec and we abort).

The cost should be very small; it's an additional trap instruction (well, two for now, since we use TrapUnreachable, but that's a different issue) for each function with landing pads; if LLVM gains support to encode "abort/terminate" info directly in LSDA like GCC does, then it'll be free. It's an additional basic block though so compile time may be worse, so I'd like a perf run.

r? @ghost
@rustbot label: F-c_unwind

### CODE REVIEW

**Comment by Amanieu** [comment]
In #92828 I am changing `TerminatorKind::Abort` to call the panic handler again to print a message instead of aborting directly with an obscure "Illegal Instruction" error. It would make sense to do the same here.

**Comment by nbdd0121** [comment]
I am a bit concerned about code bloat if this was a call to another function instead.

**Comment by nbdd0121** [comment]
I have updated the code to call `panic_no_unwind` instead. The perf isn't much different from the trap version.

I removed the `#[track_caller]` on `panic_no_unwind` because:
* The call is always implicit, so the location can only point to the function but not detailed lines;
* The function that cause unwind usually matters more than the function that forbids unwinding;
* For this function to be called, an unwind/panic has already happened, and thus should already have backtrace info;
* The backtrace by `panic_no_unwind` is sufficient to point to the function anyway.
* Most importantly, this is extremely cold path, so code size reduction is always good.

r? @Amanieu

**Comment by nbdd0121** [comment]
Placed the test in the wrong directory 🤦 fixed.

**Comment by nbdd0121** [comment]
Hmm... We now codegen differently for SEH/Itanium, so the check is failing for MSVC... Shall I just restrict the test with `ignore-msvc` or duplicate it to two copies (`ignore-msvc + only-msvc`)?

**Comment by Amanieu** [comment]
I think just disabling these tests on msvc is fine.

**Comment by nbdd0121** [comment]
I had another look at the drop test, the `CHECK-NOT` there is really just to check that there are no drop calls in between two `CHECK`s so the total number of drop calls are 6. So I just updated it to accept `(call|invoke)` for the drop call in the unwinding path.

**Comment by rylev** [comment]
@nbdd0121 @Amanieu - this produces somewhat significant regressions in compiler performance. While this is a correctness fix, the situation it's guarding against (double unwind) seems like a relatively rare occurrence. Is the performance regression here justified?

**Comment by nbdd0121** [comment]
We've discussed about the perf regression briefly in https://rust-lang.zulipchat.com/#narrow/stream/210922-project-ffi-unwind/topic/weekly.20meeting, and our conclusion is that compile time regression can't be avoided because LLVM has more work to do. This is just compile time perf regression and it wouldn't affect runtime (this is cold path of cold path).

Although currently double unwinding can be be trigger via FFI unwinding, there are some discussions about allowing Rust code to perform double unwinding too - in that case this guard will be essential.

In the short term we might have to bear with this overhead in compile time, but if we made destructors nounwind by default then the overhead should go away.

**Comment by Amanieu** [comment]
> In the short term we might have to bear with this overhead in compile time, but if we made destructors nounwind by default then the overhead should go away.

For more details see https://github.com/rust-lang/lang-team/issues/97.

File: src/test/run-make-fulldeps/foreign-double-unwind/Makefile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+-include ../tools.mk
+
+all: foo
+	$(call RUN,foo) | $(CGREP) -v unreachable
```
Comment (riking) [inline src/test/run-make-fulldeps/foreign-double-unwind/Makefile:4]: Shouldn't this be `! grep unreachable` ? -v inverts the match, so if the program writes _any_ output grep reports success.

We want instead to search for "unreachable" and fail the build if it is found.

Comment (nbdd0121) [reply src/test/run-make-fulldeps/foreign-double-unwind/Makefile:4]: `$(CGREP)` is not grep.

**Comment by RalfJung** [comment]
(Follow-up discussion [happened on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/269128-miri/topic/double_unwind_guard).)

**Pull request merged**